osu.Server.Spectator/Database/DatabaseAccess.cs

```diff
@@ -234,12 +246,32 @@ public async Task EndMatchAsync(MultiplayerRoom room)
 
         public async Task<multiplayer_playlist_item[]> GetAllPlaylistItemsAsync(long roomId)
         {
+            var connection = await getConnectionAsync();
+
             return (await connection.QueryAsync<multiplayer_playlist_item>("SELECT * FROM multiplayer_playlist_items WHERE room_id = @RoomId", new { RoomId = roomId })).ToArray();
         }
 
         public void Dispose()
         {
-            connection.Dispose();
+            openConnection?.Dispose();
+        }
+
+        private async Task<MySqlConnection> getConnectionAsync()
+        {
+            if (openConnection != null)
+                return openConnection;
+
+            string host = Environment.GetEnvironmentVariable("DB_HOST") ?? "localhost";
+            string user = Environment.GetEnvironmentVariable("DB_USER") ?? "root";
+            string port = Environment.GetEnvironmentVariable("DB_PORT") ?? "3306";
+
+            DapperExtensions.InstallDateTimeOffsetMapper();
+
+            openConnection = new MySqlConnection($"Server={host};Port={port};Database=osu;User ID={user};ConnectionTimeout=5;ConnectionReset=false;Pooling=true;");
+
+            await openConnection.OpenAsync();
```

Comment:
While this class is technically currently not used in a manner where it would need to be thread safe (fresh instances are created every time), this initialisation process is giving me the heebie-jeebies from a threading perspective. I don't think this is thread safe:

| Thread A | Thread B |
| :- | :- |
| `if (openConnection != null) // fails` | |
| | `if (openConnection != null) // fails` |
| `openConnection = new MySqlConnection(...)` | |
| `await openConnection.OpenAsync();` | |
| `return openConnection;` | |
| | `openConnection = new MySqlConnection(...)` |
| | `await openConnection.OpenAsync();` |
| | `return openConnection;` |

et voilà, there's a leaked DB connection. _Maybe_ the finaliser will close it, but this feels unnecessarily risky - although maybe it's just me being paranoid here?

Is it viable to keep `openConnection` readonly, create it in the ctor, and then only open it when needed? I'd feel much safer with that.

Reply:
Though I don't expect this class to ever be used in a situation like the above, I think the suggestion makes sense.

Reply:
I didn't see this as important, but can convert to a `Lazy` to resolve the issue.

Note that the proposal above is hard to implement because it would require locking over the `OpenAsync` call.